### ISSUE TITLE
SimClr training loss not decreasing for many many epochs in unsupervised_hello_world.ipynb

### ISSUE THREAD

**Issue by kechan** [tensorflow/similarity#233]
Just running the example verbatim from the notebook unsupervised_hello_world.ipynb for SimClr. The train loss seemed to get stuck around 6.9287. The solution has no evidence of a collapse with proj_std 0.02 for both train and val. The binary accuracy is not great at ~0.5 

Anyone ran the same in the notebook and got the same result? I wonder if there's a bug in the implementation (haven't looked in the code in detail yet). 

```
Epoch 728/800
87/87 [==============================] - ETA: 0s - loss: 6.9287 - proj_std: 0.0214binary_accuracy: 0.4910
87/87 [==============================] - 49s 561ms/step - loss: 6.9287 - proj_std: 0.0214 - val_loss: 6.9288 - val_proj_std: 0.0211 - binary_accuracy: 0.4910

```

<img width="917" alt="Screen Shot 2022-02-03 at 12 26 58 PM" src="https://user-images.githubusercontent.com/122762/152396188-be6cf9ca-f907-49dc-b07b-1cd429e59576.png">

**Comment by owenvallis** [comment]
SimClr was the one algorithm I wasn't able to repro the paper results. If I remember correctly, all the cifar results in the paper used a ResNet50 backbone that did the self-supervised pre-training on ImageNet. The current ResNet18 with cifar pre-training may not be optimal for SimClr. 

I'm pretty sure I implemented the cross-entropy loss correctly... however, it was an interesting loss as it has that rectangular shape to the loss matrix, so please me know if you find any issues. I'll take another look at it as well to double check.

**Comment by kechan** [comment]
Did some quick/dirty test today:

1) Swap in resnet50 (manually add the layer.Rescale(...) inside the model. 

    ```
   backbone = tfsim.architectures.ResNet50Sim(
        input_shape,
        weights=None,
        include_top=False,  # Take the pooling layer as the output.
        pooling="avg",      # Can change to use `gem` -> GeneralizedMeanPooling2D
        # gem_p=3.0,        # Increase the contrast between activations in the feature map.
        # preproc_mode="torch",
    )
   ```
 Summary:

```
Model: "similarity_model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 input_1 (InputLayer)        [(None, 32, 32, 3)]       0         
                                                                 
 rescale_1_over_255 (Rescali  (None, 32, 32, 3)        0         
 ng)                                                             
                                                                 
 resnet50 (Functional)       (None, None, None, 2048)  23587712  
                                                                 
 avg_pool (GlobalAveragePool  (None, 2048)             0         
 ing2D)                                                          
                                                                 
=================================================================
Total params: 23,587,712
Trainable params: 0
Non-trainable params: 23,587,712
```

And train only with a single batch of data:

```
history = contrastive_model.fit(
    train_ds.take(1),
    epochs=PRE_TRAIN_EPOCHS,
    steps_per_epoch=1,
)
```

This should overfit and reduce train loss to ~0 (assuming the SimClr loss is designed that way, need to reread paper). But it appears the loss is stuck around 6.9294... 

I will later come around and see what I can find. But it seems SimClr may be a bit tough to utilize for now (at least for me).

**Comment by kechan** [comment]
Update: I just realized the loss may not still steadily go down to 0, due to heavy data aug during training, so the sanity test i mentioned above may not be ideal. I can probably turn the data aug off and make it deterministic, and then be sure the loss go down to a value expected from the loss function itself.

**Comment by kechan** [comment]
Update: I trained with 512

```
    train_ds.take(1),
    epochs=PRE_TRAIN_EPOCHS,
    steps_per_epoch=1,
```

The loss stayed fixed at 6.9286. So something is a bit fishy, just don't know where yet.

**Comment by owenvallis** [comment]
Hi Kechan,

Thanks so much for looking into this. I'm hoping to dig into this next week if I can. I looked over the code and I remember how I was thinking about things when I implemented this. Basically, I see the loss as cross entropy over two pairwise matrices, view_a x view_b and view_a x view_a, on which we do a column wise concatenate. In a perfect model we have 1.0 along the diagonal of the first matrix and zeros everywhere else. This is why I tried to mask out the diagonal of the second matrix to remove the self similarity values. My assumption is that we can then just treat the concatenated matrices as basic cross entropy at that point. I think a good next step will be to compute the loss step by step as described in the paper and then compare that against our current cross entropy implementation. I'll try and test it out in a Colab next week.

**Comment by kechan** [comment]
@owenvallis Thanks. 

I think I may spotted at least one error in class SimCLRLoss:

```
   if self.use_hidden_norm:
            za = tf.math.l2_normalize(za)
            zb = tf.math.l2_normalize(zb)
```

The l2_norm should likely be acted only on axis=-1, and not across batch dimension. The default for axis is actually None for tf.math.l2_normalize(...). I think TF may have inconsistent convention for axis default, i do see other function with axis=-1 as default. So this can be quite confusing. 

This is after a wide goose chase after other dead ends, so i will take a break and come to re-test everything with the above corrected, to see if it makes any progress. But I highly suspect this is the main cause. I looked through other code in loss and they do match the overall logic from the paper.

**Comment by kechan** [comment]
Update: Changed in SimCLRLoss

```
if self.use_hidden_norm:
  za = tf.math.l2_normalize(za, axis=-1)
  zb = tf.math.l2_normalize(zb, axis=-1)
```

Testing on only 1 batch of data with resnet50 (full trainable):
```
tf.keras.optimizers.SGD(learning_rate=1e-4, momentum=0.9)
#
train_ds.take(1),
epochs=PRE_TRAIN_EPOCHS,
steps_per_epoch=1,
```

The train loss is now making clear progress decreasing, at least compared with what's before. 

<img width="590" alt="Screen Shot 2022-02-21 at 1 43 24 PM" src="https://user-images.githubusercontent.com/122762/155011585-1352fedc-9947-448b-a29a-5a163ed3151f.png">

This decrease is actually quite noisy and probably make sense since it is a more difficult task than the standard classification objective with a single target. 

I think my next test step is to revert everything back to original notebook setup, except with using ResNet50 backbone, instead of ResNet18.

**Comment by kechan** [comment]
Update: Reverting all setup back to original with only backbone change: 

```
backbone = tfsim.architectures.ResNet50Sim(
        input_shape,
        weights=None,
        include_top=False,  # Take the pooling layer as the output.
        pooling="avg",      # Can change to use `gem` -> GeneralizedMeanPooling2D
        trainable="full",
        # gem_p=3.0,        # Increase the contrast between activations in the feature map.
        # preproc_mode="torch",
    )
```

I also manually added: 
   `x = layers.Rescaling(1./255, name='rescale_1_over_255')(x) `

to be compatible with ResNet18Sim so nothing on the dataset preparation needed to be changed in notebook. 

After only 2 epoch, my loss is already lowered to 6.0, arguably already better than my 1st trial at the notebook. 

I am filing an separate issue/bug for this, and will use this thread to keep reporting my progress on CIFAR10/ResNet50Sim training.

**Comment by owenvallis** [comment]
Thanks for catching this! I'll push a patch to the master branch and link it in the other issue.

**Comment by kechan** [comment]
Here's the training result:
<img width="876" alt="Screen Shot 2022-02-21 at 10 01 57 PM" src="https://user-images.githubusercontent.com/122762/155055451-d3bc2ce3-db89-41de-9e8c-889fda4dd599.png">

Noticeably improved after the l2norm fix. But superficially, it doesn't perform in terms of binary_accuracy as Simsiam? 

Not sure how this result should be compared, and if this actually reproduces fairly the SimClr paper's declared result. 

But probably can use this now for downstream fine tuning task in our own projects.

**Comment by owenvallis** [comment]
That's great to see! I'm also training SimCLR with the same CIFAR / ResNet18 setup that I have in the notebook for SimSiam, and I'm at epoch 200 right now with a binary acc of roughly 0.72. This seems a little higher than I was seeing for SimSiam at the same point, but I'll have to wait and see what we get to at epoch 800. I'll do the same for Barlow Twins as well, and this will hopefully let us compare all three algos.

**Comment by kechan** [comment]
@owenvallis For CIFAR10, it may actually make sense if ResNet18 outperforms, assuming I didn't mess up my mod of ResNet50Sim. ResNet50 may shine only with "imagenet" size of dataset. In my previous experience for single target classification, I have seen huge model can suffer from overfitting for small/medium size dataset. So if true, then we may be seeing the same thing for contrastive model. But I also vaguely remember in the paper, they mention their method benefit from large batch size and huge data...

**Comment by owenvallis** [comment]
Here's the SimCLR results using the ResNet18 backbone.

![image](https://user-images.githubusercontent.com/916864/155186482-bbe990bb-295d-403b-8ee4-0d34c13de278.png)

This is similar to the SimSiam results, although the final acc is a little lower. Seems like we might be able to increase the LR  a little, but this seems like a reasonable result.

I should have the Barlow Twins results today as well, and I'll push the normalization change as a patch to the master branch.

**Comment by owenvallis** [comment]
Actually, the Barlow Twins loss is fine. I'm just updating the helper method name to make it clear that we are standardizing the columns instead of normalizing them.

**Comment by owenvallis** [comment]
So I ran some checks and I think we are good to go now that the normalization is fixed. I'll add these as tests in the PR.

![image](https://user-images.githubusercontent.com/916864/155204620-99bc09d5-019b-4c7a-a275-6721d53c29a0.png)

**Comment by owenvallis** [comment]
Thanks again for bringing this to our attention. Closing this issue as well as we pushed the fix in 0.15.4

**Comment by ManafMukred** [comment]
> Here's the training result: <img alt="Screen Shot 2022-02-21 at 10 01 57 PM" width="876" src="https://user-images.githubusercontent.com/122762/155055451-d3bc2ce3-db89-41de-9e8c-889fda4dd599.png">
> 
> Noticeably improved after the l2norm fix. But superficially, it doesn't perform in terms of binary_accuracy as well as Simsiam?
> 
> Not sure how this result should be compared, and if this actually reproduces fairly the SimClr paper's declared result.
> 
> But probably can use this now for downstream fine tuning task in our own projects.

@kechan  I tried to redo this on resnet50 ... but the error still 6 and never inproves
